pkg/gatewayserver/io/basicstationlns/basicstationlns_test.go

```diff
@@ -60,13 +55,79 @@ var (
 func eui64Ptr(eui types.EUI64) *types.EUI64 { return &eui }
 
 func TestAuthentication(t *testing.T) {
-	// TODO: Test authentication. We're gonna provision authentication tokens, which may be API keys.
-	// https://github.com/TheThingsNetwork/lorawan-stack/issues/558
+	ctx := log.NewContext(test.Context(), test.GetLogger(t))
+
+	c := component.MustNew(test.GetLogger(t), &component.Config{
+		ServiceBase: config.ServiceBase{
+			HTTP: config.HTTP{
+				Listen: ":8100",
+			},
+		},
+	})
+	gs := mock.NewServer()
+	srv := New(ctx, gs)
+	c.RegisterWeb(srv)
+	test.Must(nil, c.Start())
+	defer c.Close()
+
+	gs.RegisterGateway(ctx, registeredGatewayID, &registeredGateway, registeredGatewayToken)
+
+	for _, tc := range []struct {
+		Name           string
+		GatewayID      string
+		AuthToken      string
+		ErrorAssertion func(err error) bool
+	}{
+		{
+			Name:           "RegisteredGatewayAndValidKey",
+			GatewayID:      registeredGatewayID.GatewayID,
+			AuthToken:      registeredGatewayToken,
+			ErrorAssertion: nil,
+		},
+		{
+			Name:      "RegisteredGatewayAndInValidKey",
+			GatewayID: registeredGatewayID.GatewayID,
+			AuthToken: "invalidToken",
+			ErrorAssertion: func(err error) bool {
+				return err == websocket.ErrBadHandshake
+			},
+		},
+		{
+			Name:           "RegisteredGatewayAndNoKey",
+			GatewayID:      registeredGatewayID.GatewayID,
+			ErrorAssertion: nil,
+		},
+		{
+			Name:      "UnregisteredGateway",
+			GatewayID: "eui-1122334455667788",
+			AuthToken: registeredGatewayToken,
+			ErrorAssertion: func(err error) bool {
+				return err == websocket.ErrBadHandshake
+			},
+		},
+	} {
+		t.Run(fmt.Sprintf("%s", tc.Name), func(t *testing.T) {
+			a := assertions.New(t)
+			h := http.Header{}
+			h.Set("Authorization", tc.AuthToken)
+			conn, _, err := websocket.DefaultDialer.Dial("ws://localhost:8100/api/v3/gs/io/basicstation/traffic/"+tc.GatewayID, h)
+			if err != nil {
+				if tc.ErrorAssertion == nil || !a.So(tc.ErrorAssertion(err), should.BeTrue) {
+					t.Fatalf("Unexpected error: %v", err)
+				}
+			} else if tc.ErrorAssertion != nil {
+				t.Fatalf("Expected error")
+			}
+			if conn != nil {
```

Comment:
`conn` is undefined if `err != nil`; consider `defer conn.Close()` right after `if err != nil`

Reply:
How can I call the `Close()` method if the object is `nil`?

Reply:
>right after `if err != nil`

That is where `conn != nil`, i.e. L115

Reply:
Not addressed